src/main/scala/com/salesforce/spark/carrot/RabbitUtils.scala

```diff
@@ -0,0 +1,206 @@
+ *
+ * See ConnectionFactory for details on RabbitMQ specific parameters.
+ *
+ * @param rmqConfig RMQConfig that defines RMQ behavior
+ */
+class ConnectionFactoryWithFailover(rmqConfig: RMQConfig) extends ConnectionFactory() {
+  var ifPrimaryHost = true
+  this.setHost(rmqConfig.hostname)
+  this.setPort(rmqConfig.port)
+  rmqConfig.username.foreach(this.setUsername _)
+  rmqConfig.password.foreach(this.setPassword _)
+
+  def toggleHost = {
+    if (ifPrimaryHost) this.setHost(rmqConfig.alternativeHost)
+    else if (!ifPrimaryHost) this.setHost(rmqConfig.hostname)
+
+    ifPrimaryHost = !ifPrimaryHost
+  }
+
+  /**
+   * This is an exponential delay on the calls to RabbitMQ.  If a machine is unresponsive, exponentially wait.
+   * Also swap between this host and the alternative host if there is one.
+   */
+  override def newConnection: Connection = {
+    val connectionExponentialDelay = new ExponentialDelay(1000, 600000)
+    while (true) {
+      try {
+        val connection = super.newConnection
+        connectionExponentialDelay.reset()
+        return connection
+      } catch {
+        case ioe: ConnectException => {
+          connectionExponentialDelay.pulse()
+          toggleHost
+        }
+        case e: Exception => throw e
+      }
+    }
+    throw new IllegalStateException("This code path should be inaccessible.")
+  }
+}
+
+class RMQReceiver[T: ClassTag, U <: Decoder[_] : ClassTag]
+(
+  rmqConfig: RMQConfig,
+  @transient storageLevel: StorageLevel
+  )
+  extends Receiver[T](storageLevel) with Logging with Serializable {
+
+  def onStart {
+    // Start the thread that receives data over a connection
+    new Thread("Socket Receiver") {
+      override def run {
+        receive
+      }
+    }.start
+  }
+
+  def onStop {
+    // There is nothing much to do as the thread calling receive()
+    // is designed to stop by itself isStopped() returns false
+  }
+
+  /**
+   * Create a connection from an rmqConfig
+   *
+   * @param rmqConfig
+   * @return rabbitMQ connection object
+   */
+  private def connectRabbitMq(rmqConfig: RMQConfig) = new ConnectionFactoryWithFailover(rmqConfig).newConnection
+
+  /** Create a socket connection and receive data until receiver is stopped */
+  private def receive {
+
+    //Handle a decoder that has an implicit ClassTag; failing this, try to use a decoder with an empty constructor
+    val valueDecoder = Try(classTag[U].runtimeClass.getConstructor(classOf[ClassTag[T]]))
+      .map(_.newInstance(classTag[T]).asInstanceOf[Decoder[T]])
+      .getOrElse(classTag[U].runtimeClass.newInstance.asInstanceOf[Decoder[T]])
+
+    val connection = connectRabbitMq(rmqConfig)
+    val channel = connection.createChannel
+
+    logInfo(s"Beginning processing queue: ${rmqConfig.queue}")
+
+    val consumer = new QueueingConsumer(channel)
+    channel.basicConsume(rmqConfig.queue, false, consumer)
+
+    val delayer = new ExponentialDelay
```

Comment:
@suryasev it's better to explicitly specify `ExponentialDelay` parameters, rather than using defaults